Improve ztest macros

barnas-michal opened the pull request:
These changes removes the requirement for sending the NULL as last parameter if the message is not used in zassert_* and zassume_* macros.

If the default message is descriptive enough, there's no need to provide the additional one, but it was required to send NULL as parameter to the macros. It will no longer be needed after this change.

barnas-michal committed:
ztest: improve comments for ztest assert macros

Move params in comment about zassert and zassume macros.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
ztest: prepare zassert and zassume macro to be used without message

This commit adds intermediate macros that will allow to use the
zassert and zassume macros without setting the msg parameter as NULL
if it isn't used.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

barnas-michal committed:
ztest: remove msg parameter from zassert_* and zassume_* macros

This commit removes the requirement for msg parameter in zassert_*
and zassume_* macros. It will allow to use them without sending NULL
as the last parameter.

The default messages look like:

void* beauty_ptr = NULL;
zassert_not_null(beauty_ptr);
=>
Assertion failed at main.c:20: suite_test: beauty_ptr is NULL

void* ugly_ptr = (void*)0xbaadf00d;
zassume_is_null(ugly_ptr);
=>
Assumption failed at main.c:23: suite_test: ugly_ptr is not NULL

int apple = 3;
int tomato = 9;
zassume_equal(apple, tomato);
=>
Assumption failed at main.c:27: suite_test: apple not equal to tomato

Signed-off-by: Michał Barnaś <mb@semihalf.com>

yperess reviewed (changes requested):
Can you update some tests and also include a sample error message if none was specified?

barnas-michal committed:
ztest: remove the obsolete NULL appended to zassert macros

This commit removes the usage of NULL parameter as message in
zassert_* macros after making it optional

Signed-off-by: Michał Barnaś <mb@semihalf.com>

aaronemassey commented on subsys/testsuite/ztest/include/zephyr/ztest_assert.h:
```diff
@@ -141,6 +141,19 @@ static inline bool z_zassume(bool cond, const char *default_msg, const char *fil
+ * You probably don't need to call this macro directly. You should
+ * instead use zassume_{condition} macros below.
+ *
+ * @param cond Condition to check
```
This will conflict with https://github.com/zephyrproject-rtos/zephyr/pull/48667/files which I think will be merged today. So may want to hold off and rebase.

semihalf-barnas-michal replied:
Good to know, I will wait

aaronemassey reviewed (dismissed):
LGTM, but should wait to rebase off those doc changes I'm about to merge I think.

barnas-michal committed:
ztest: improve some tests

This commit changes some tests from using zassert_equal to validate
the pointers to using the zassert_is_null and zassert_not_null.

Signed-off-by: Michał Barnaś <mb@semihalf.com>

semihalf-barnas-michal commented:
> Can you update some tests and also include a sample error message if none was specified?
Sure, I've added it also to the commit message of the one that updates the zassert_* macros.

```
void* beauty_ptr = NULL;
zassert_not_null(beauty_ptr);
```
=>
`Assertion failed at main.c:20: suite_test: beauty_ptr is NULL`

```
void* ugly_ptr = (void*)0xbaadf00d;
zassume_is_null(ugly_ptr);
```
=>
`Assumption failed at main.c:23: suite_test: ugly_ptr is not NULL`

```
int apple = 3;
int tomato = 9;
zassume_equal(apple, tomato);
```
=>
`Assumption failed at main.c:27: suite_test: apple not equal to tomato`

stephanosio commented on tests/unit/cbprintf/main.c:
```diff
@@ -584,7 +584,7 @@ static void test_d_length(void)
 	rc = rawprf("/%Ld/", max);
 	zassert_equal(rc, 5, "len %d", rc);
-	zassert_equal(strncmp("/%Ld/", buf, rc), 0, NULL);
+	zassert_equal(strncmp("/%lld/", buf, rc), 0);
```
Please leave `%Ld` here as is. From what I can see, this is intentionally testing an invalid format specifier.

semihalf-barnas-michal replied:
Yes, you're right. I didn't revert this patch yet, because I am looking for a help with suppressing the checkpatch warning about this format.

yperess reviewed (changes requested):
It would be nice to have actual framework tests for this. I think you could add a test under `tests/ztest/assert` that do something like:
```
ZTEST_EXPECT_FAIL(asserts, test_default_assert_failure_message)
ZTEST(asserts, test_default_assert_failure_message)
{
  zassert_false(true);
}
```

Then in that test's testcase.yaml you can use the harness_config like:
```
harness_config:
  type: multi_line
  ordered: false
  regex:
    - "<the default error message>"
```

(I believe that this should all work)

yperess reviewed (dismissed):
Spoke with Michael offline and he's actively working on being able to use the console test harness with ztest. Having that will allow us to create tests that assert on the error message. Since these tests never existed, I don't see a reason to block this PR as it makes the assert and assume API much better.

semihalf-barnas-michal commented:
> It would be nice to have actual framework tests for this
I will do them in separate PR, since it requires to fix the regexes verification in the ztest framework.

This PR has compliance check failing due to the test that checks for handling of invalid printf format string. This is a false-positive check and should be ignored.
`WARNING:PRINTF_L: %Ld is non-standard C, use %lld`

andyross reviewed (dismissed):
Not sure if I ever reviewed this one or not, but I love the feature.  Care to play the same trick with __ASSERT() so we can eliminate ___ASSERT_NO_MSG()? :)

semihalf-barnas-michal commented:
> Not sure if I ever reviewed this one or not, but I love the feature. Care to play the same trick with __ASSERT() so we can eliminate ___ASSERT_NO_MSG()? :)

Sure, I can take a look at this later :)

fabiobaltieri commented:
CI fail is unrelated:

````
#15420: FILE: tests/unit/cbprintf/main.c:527:
+	zassert_equal(strncmp("/%Ld/", buf, rc), 0);
```

That'll have to be bypassed.

henrikbrixandersen commented on subsys/testsuite/ztest/include/zephyr/ztest_assert.h:
```diff
@@ -143,6 +143,12 @@ static inline bool z_zassume(bool cond, const char *default_msg, const char *fil
+	_zassert_base(cond, default_msg, msg, ##__VA_ARGS__)
+
+#define zassert(cond, default_msg, ...)                                                            \
+	_zassert_va(cond, default_msg, COND_CODE_1(__VA_OPT__(1), (__VA_ARGS__), (NULL)))
```
Is `__VA_OPT__` supported by all the toolchains supported by Zephyr? /cc @stephanosio

stephanosio replied:
For one, `__VA_OPT__` is non-standard and a GNU extension.

I am not sure if it is supported by the XCC, which is based on GCC 4.x. Can you confirm? @nashif @dcpleung @laurenmurphyx64

nashif replied:
looks like it is.

fabiobaltieri commented:
@stephanosio can you manually merge this?

stephanosio commented:
> @stephanosio can you manually merge this?

uhh .. looks like we have a conflict now.

stephanosio commented:
@semihalf-barnas-michal can you please rebase?

fabiobaltieri reviewed (approved):
@stephanosio this should be good to go now.

semihalf-barnas-michal commented:
> can you please rebase?

@stephanosio : Sure, done. The only warning is a false-positive against the sprintf format test.

nashif commented:
> @stephanosio this should be good to go now.

it is only good to go when reviewed by assignee, whoever that might be :-)
